[WIP] Add TEP to disable default creation of RBAC resources

savitaashture opened the pull request:
/cc @vdemeester @ppitonak @nikhil-thomas

vdemeester commented on enhancements/teps/0001-option-to-disable-default-injection-of-serviceaccount-rbac.md:
```diff
@@ -0,0 +1,114 @@
+
+## Motivation
+
+OpenShift Pipelines operator will create a RBAC resources ((`pipeline`)ServiceAccount, RoleBinding, SCCRoleBinding, CABundlesConfigMap and Subjects to openshift-pipelines-clusterinterceptors) on all namespace when installed.
```
`s/will create/creates` as it is how it works today.

vdemeester reviewed (commented):
We need to document the risks/potential issues with this. One here would be : what happens to the default SA (`pipeline`) in that case ? How does the operator handles this when its creation is disabled ? cluster-wide, or by namespace ?

savitaashture commented on enhancements/teps/0001-option-to-disable-default-injection-of-serviceaccount-rbac.md:
```diff
@@ -0,0 +1,114 @@
+metadata:
+  name: config
+spec:
+  disable-rbac-clusterwide: true
```
So as per the discussion from Slack https://coreos.slack.com/archives/CG5GV6CJD/p1632132969004700 

shall we go ahead with adding new filed `params` 

```
apiVersion: operator.tekton.dev/v1alpha1
kind: TektonConfig
metadata:
  name: config
spec:
  params:
  - name: OpenShiftPipelinesifyNamespaces
    value: true
```
OR
```
apiVersion: operator.tekton.dev/v1alpha1
kind: TektonConfig
metadata:
  name: config
spec:
  params:
  - name: ConfigureRbacResources
    value: true
```

/cc @nikhil-thomas @sm43 @vdemeester

vdemeester replied:
We need to weight what we are disabling and what could it mean in the future. Also, `OpenShiftPipelinesifyNamespaces` doesn't really mean anything to the user, so I wouldn't go that way.

vdemeester commented on enhancements/teps/0001-option-to-disable-default-injection-of-serviceaccount-rbac.md:
```diff
@@ -0,0 +1,114 @@
+status: implementable
+---
+
+# TEP-0001: Option to disable dafault creation of service account  and rbac resources
```
"On OpenShift" ? As, today, we are only having this on OpenShift platform.

vdemeester commented on enhancements/teps/0001-option-to-disable-default-injection-of-serviceaccount-rbac.md:
```diff
@@ -0,0 +1,114 @@
+## Proposal
+
+Installation of OpenShift Pipelines operator by default create RBAC resources on all namespaces.
+cluster admin will get permission to disable RBAC resource creation at cluster level using `TektonConfig CR` and at namespace level using `label`. 
```
"Cluster admin should have the permission to …"

vdemeester commented on enhancements/teps/0001-option-to-disable-default-injection-of-serviceaccount-rbac.md:
```diff
@@ -0,0 +1,114 @@
+cluster admin will get permission to disable RBAC resource creation at cluster level using `TektonConfig CR` and at namespace level using `label`. 
+   
+### User Stories
+Some customers have reported that SCCRolebinding `pipelines-scc-rolebinding` can be seen as a security issue which has **RunAsAny** among other things.
```
A user story should be something more like `As a cluster-admin, I want to be able to …`

vdemeester reviewed (changes requested):
Having the per-namespace options is adding more questions than anything as it stands (before [TEP-0080](https://github.com/tektoncd/community/pull/506) is in). How default SA in configuration would work if we disable or enable the creation of this in some namespace and not in others (without having a way to enable per-namespace configurations).

So I would suggest we reduce this design to just enable/disable cluster-wide, and create a follow-up epic (and design) for per-namespace. Then we just need to make sure that, if the cluster-admin disables this automatic SA creation, we also disable the default SA configuration for pipeline (so that it doesn't try to use a non-existent SA).

nikhil-thomas commented:
> Having the per-namespace options is adding more questions than anything as it stands (before [TEP-0080](https://github.com/tektoncd/community/pull/506) is in). How default SA in configuration would work if we disable or enable the creation of this in some namespace and not in others (without having a way to enable per-namespace configurations).
> 
> So I would suggest we reduce this design to just enable/disable cluster-wide, and create a follow-up epic (and design) for per-namespace. Then we just need to make sure that, if the cluster-admin disables this automatic SA creation, we also disable the default SA configuration for pipeline (so that it doesn't try to use a non-existent SA).

I agree with this. Narrowing our scope to Enabling/Disabling the feature clusterwide wiil be a good start. 


In fact, in the current design  im not 100% sure about, how our user's interpret the toggle nature of the namespace annotatiton when we set / unset the clusterwide configuration

savitaashture commented:
> Having the per-namespace options is adding more questions than anything as it stands (before [TEP-0080](https://github.com/tektoncd/community/pull/506) is in). How default SA in configuration would work if we disable or enable the creation of this in some namespace and not in others (without having a way to enable per-namespace configurations).
> 
> So I would suggest we reduce this design to just enable/disable cluster-wide, and create a follow-up epic (and design) for per-namespace. Then we just need to make sure that, if the cluster-admin disables this automatic SA creation, we also disable the default SA configuration for pipeline (so that it doesn't try to use a non-existent SA).

Yes totally agree :+1: 
Lets go with step by step and based on user experience we can keep on enhancing this feature.

I have updated TEP based on this comment and updated design to enable/disable cluster-wide.

savitaashture commented on enhancements/teps/0001-option-to-disable-default-injection-of-serviceaccount-rbac.md:
```diff
@@ -0,0 +1,124 @@
+metadata:
+  name: config
+spec:
+  params:
+  - name: disableRbacResourcesForOpenshift
+    value: true
```
@vdemeester @nikhil-thomas @sm43 how about this?

vdemeester replied:
So a user of the OpenShift platform (and OpenShift Pipelines), will have to set this too, which might be weird (aka why `disableRbacResourcesForOpenShift` instead of just `disableRbacResources` ?).

Because we have this in `params`, and thus not *typed*, any reason for having `ForOpenShift` suffix there ? on Kubernetes setting this would/should be a no-op (at least today).

savitaashture replied:
>Because we have this in params, and thus not typed, any reason for having ForOpenShift suffix there ? on Kubernetes setting this would/should be a no-op (at least today).

Yes on Kubernetes setting this leads to no-op
But for clarity i have used `ForOpenShift` suffix

But it make sense to use `CreateRbacResource` 

>So a user of the OpenShift platform (and OpenShift Pipelines), will have to set this too, which might be weird

Ya thats what i assume i mean if user doesn't set that key then how do we decide disabling auto creation :thinking:

savitaashture commented:
@nikhil-thomas @vdemeester @sm43 i have updated TEP please take a look :bowing_woman:

vdemeester commented on enhancements/teps/0001-option-to-disable-default-injection-of-serviceaccount-rbac.md:
```diff
@@ -0,0 +1,124 @@
+  name: config
+spec:
+  params:
+  - name: disableRbacResourcesForOpenshift
```
Could be `CreateRbacResource` if we don't want to have double negative or something 👼🏼

savitaashture replied:
Yes
```
apiVersion: operator.tekton.dev/v1alpha1
kind: TektonConfig
metadata:
  name: config
spec:
  params:
  - name: createRbacResource
    value: true
```
will auto create RBAC resources

```
apiVersion: operator.tekton.dev/v1alpha1
kind: TektonConfig
metadata:
  name: config
spec:
  params:
  - name: createRbacResource
    value: false
```
will disable auto creation of RBAC resources 


Shall i go ahead with this ?
/cc @nikhil-thomas @vdemeester

sm43 commented on enhancements/teps/0001-option-to-disable-default-injection-of-serviceaccount-rbac.md:
```diff
@@ -0,0 +1,125 @@
+---
+title: Option to disable dafault creation of service account and rbac resources
```
```suggestion
title: Option to disable default creation of service account and rbac resources
```

sm43 commented:
/lgtm
